nemo/core/classes/modelPT.py

```diff
@@ -555,8 +557,22 @@ def setup_optimization(self, optim_config: Optional[Union[DictConfig, Dict]] = N
                     raise e
 
         else:
+            # Pop arguments for Megatron AMP-O2 training
+            megatron_amp_o2 = optimizer_args.pop('megatron_amp_o2', False)
+            fp32_grad_accum = optimizer_args.pop('fp32_grad_accum', False)
+            contiguous_grad_bucket = optimizer_args.pop('contiguous_grad_bucket', False)
+            async_grad_allreduce = optimizer_args.pop('async_grad_allreduce', False)
+
             optimizer = optim.get_optimizer(optimizer_name)
             optimizer = optimizer(self.parameters(), **optimizer_args)
+            if megatron_amp_o2:
+                # Wrap the baseline optimizer with an optimizer class with master parameters
+                optimizer = MasterOptimizerWrapper(
+                    optimizer,
+                    fp32_grad_accum=fp32_grad_accum,
+                    contiguous_grad_bucket=contiguous_grad_bucket,
+                    async_grad_allreduce=async_grad_allreduce,
+                )
```

Comment:
I'm wondering if we can make this less invasive. Can we create a hook that will let NeMo models customize `optimizer`. I did something similar here: https://github.com/NVIDIA/NeMo/commit/d5b2011df6426efd8ee874fa5b299c794f17b02f

Reply:
This should not be part of ModelPT really. It's not general to even a domain, but to exactly one model - Megatron. The hooks here go all over the Megatron codebase to wrap modules to force cast to fp16 etc which makes this unusable in every other domain unless it supports such operations. 

As Eric said, please move this to some form of pre/post callback function that is called inside configure_optimizers() after finishing setup_optimization(). Then NLP model can override those specific methods there. Of course, these arguments should no longer be part of model.optim either - they aren't for any optimizer but specific to supported optimizer that Megatron uses.

Reply:
Thanks for the feecback. I will change the structure as suggested.